src/CallManager.cpp

```diff
@@ -0,0 +1,320 @@
+  if (callid_ == callCandidatesEvent.content.call_id) {
+    if (onActiveCall())
+      session_.acceptICECandidates(callCandidatesEvent.content.candidates);
+    else {
+      // CallInvite has been received and we're awaiting localUser to accept or reject the call
+      for (const auto &c : callCandidatesEvent.content.candidates)
+        remoteICECandidates_.push_back(c);
+    }
+  }
+}
+
+void
+CallManager::handleEvent(const RoomEvent<CallAnswer> &callAnswerEvent)
+{
+  nhlog::ui()->debug("CallManager::incoming CallAnswer from {} with id {}", callAnswerEvent.sender, callAnswerEvent.content.call_id);
+  if (onActiveCall() && callid_ == callAnswerEvent.content.call_id) {
+    stopRingtone();
+    if (!session_.acceptAnswer(callAnswerEvent.content.sdp)) {
+      emit ChatPage::instance()->showNotification("Problem setting up call");
+      hangUp();
+    }
+  }
+}
+
+void
+CallManager::handleEvent(const RoomEvent<CallHangUp> &callHangUpEvent)
+{
+  nhlog::ui()->debug("CallManager::incoming CallHangUp from {} with id {}", callHangUpEvent.sender, callHangUpEvent.content.call_id);
+  if (callid_ == callHangUpEvent.content.call_id) {
+    MainWindow::instance()->hideOverlay();
+    endCall();
+  }
+}
+
+void
+CallManager::generateCallID()
+{
+  using namespace std::chrono;
+  uint64_t ms = duration_cast<milliseconds>(system_clock::now().time_since_epoch()).count();
+  callid_ = "c" + std::to_string(ms);
+}
+
+void
+CallManager::endCall()
+{
+  stopRingtone();
+  session_.end();
+  roomid_.clear();
+  callid_.clear();
+  remoteICECandidates_.clear();
+}
+
+void
+CallManager::retrieveTurnServer()
+{
+  http::client()->get_turn_server(
+    [this](const mtx::responses::TurnServer &res, mtx::http::RequestErr err) {
+        if (err) {
+            turnServerTimer_.setInterval(5000);
+            return;
+        }
+        emit turnServerRetrieved(res);
+  });
+}
+
+void
+CallManager::setTurnServers()
+{
+  // gstreamer expects: turn(s)://username:password@host:port?transport=udp(tcp)
+  // where username and password are percent-encoded
+  std::vector<std::string> uris;
+  for (const auto &uri : turnServer_.uris) {
+    if (auto c = uri.find(':'); c == std::string::npos) {
+      nhlog::ui()->error("Invalid TURN server uri: {}", uri);
+      continue;
+    }
+    else {
+      std::string scheme = std::string(uri, 0, c);
+      if (scheme != "turn" && scheme != "turns") {
+        nhlog::ui()->error("Invalid TURN server uri: {}", uri);
+        continue;
+      }
+
+      QString encodedUri = QString::fromStdString(scheme) + "://" + 
+                           QUrl::toPercentEncoding(QString::fromStdString(turnServer_.username)) + ":" +
+                           QUrl::toPercentEncoding(QString::fromStdString(turnServer_.password)) + "@" +
+                           QString::fromStdString(std::string(uri, ++c));
+      uris.push_back(encodedUri.toStdString());
+    }
+  }
+  if (!uris.empty())
+    session_.setTurnServers(uris);
```

Comment:
There is some issues with turn servers still on my end. Calling my phone, when it is not in my LAN doesn't seem to connect, while it works the other way around. May not be our fault though, could also be in gstreamer or just my turn server.

Reply:
Turn server support is dependent on the GStreamer uri parsing fix. In your coturn logs you'll likely see usernames like <1595006316> rather than user <1595006316:mxid> since the parsing doesn't handle the embedded colon properly.

Reply:
No, I applied the fix and rebuilt gstreamer. It does auth correctly with my turn server now. I guess there is simply still a bug in gstreamer...